### PR TITLE
feat: force option to enble retries of failed payments

### PR DESCRIPTION
**Pull request by dkoo**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Forces the WooCommerce Subscriptions option to "Retry Failed Payments" to ON unless the `NEWSPACK_PREVENT_WC_ALLOW_FAILED_PAYMENT_RETRIES_OVERRIDE` constant is enabled in wp-config.php.

### How to test the changes in this Pull Request:

1. Check out this branch.
2. Similarly to #2784, go to **WP admin > WooCommerce > Settings > Subscriptions** and test that "Retry Failed Payments" is enabled by default and can't be turned off.
3. Optionally, find an existing subscription ID on your site and update or delete the `_stripe_source_id` meta value to force the next renewal payment to fail (`wp post meta update <subscription_id> _stripe_source_id fake_token`) then trigger a manual renewal for that subscription in WP admin and confirm that it enters renewal retry mode (note the "**Renewal Payment Retry:** In 12 hours" in the Schedule panel):

Screenshot 2023-12-13 at 7 27 53 PM

4. Add the `NEWSPACK_PREVENT_WC_ALLOW_FAILED_PAYMENT_RETRIES_OVERRIDE` constant to wp-config.php and confirm you can update the option as usual.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206145382718791

### CODE REVIEW

**Review by leogermani — CHANGES_REQUESTED** [review]
I believe you want to update the comments on the `force_allow_subscription_switching` callback method, where it describes exactly what that callback is affecting.

Also, is it something that will always be tied along those other options? My question is, does it make sense to use the same constant or should it be a constant of its own? Feels like something different

**Comment by dkoo** [comment]
> Also, is it something that will always be tied along those other options? My question is, does it make sense to use the same constant or should it be a constant of its own? Feels like something different

Good call. I've separated the functions in in b8adb3e0 and updated the testing instructions with the new constant name.

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php [hunk 1]
```diff
@@ -1167,6 +1169,25 @@ public static function force_allow_subscription_switching( $can_switch, $option_
...
+	 * See: https://woo.com/document/subscriptions/failed-payment-retry/
+	 * 
+	 * @param bool   $should_retry Whether WooCommerce should automatically retry failed payments.
+	 * @param string $option_name The name of the option.
+	 *
+	 * @return string Option value.
+	 */
+	public static function force_allow_failed_payment_retry( $should_retry, $option_name ) {
+		if ( defined( 'NEWSPACK_PREVENT_WC_ALLOW_FAILED_PAYMENT_RETRIES_OVERRIDE' ) && NEWSPACK_PREVENT_WC_ALLOW_FAILED_PAYMENT_RETRIES_OVERRIDE ) {
+			return $can_switch;
```
Comment (leogermani) [inline includes/reader-revenue/woocommerce/class-woocommerce-connection.php:1184]:
```suggestion
			return $should_retry;
```

Comment (leogermani) [reply includes/reader-revenue/woocommerce/class-woocommerce-connection.php]: nit. if we don't need `option_name` here, we can remove it from the signature and the `2` from `add_filter`

Comment (dkoo) [reply includes/reader-revenue/woocommerce/class-woocommerce-connection.php]: Yes, I removed the unused arg in 4d99f80 but forgot to remove it from the action callback definition. Fixed that in 36b10f1a

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php [hunk 2]
```diff
@@ -45,8 +45,10 @@ public static function init() {
 		\add_action( 'admin_init', [ __CLASS__, 'disable_woocommerce_setup' ] );
 		\add_filter( 'option_woocommerce_subscriptions_allow_switching', [ __CLASS__, 'force_allow_subscription_switching' ], 10, 2 );
 		\add_filter( 'option_woocommerce_subscriptions_allow_switching_nyp_price', [ __CLASS__, 'force_allow_subscription_switching' ], 10, 2 );
+		\add_filter( 'option_woocommerce_subscriptions_enable_retry', [ __CLASS__, 'force_allow_failed_payment_retry' ], 10 );
```
Comment (leogermani) [inline includes/reader-revenue/woocommerce/class-woocommerce-connection.php:48]:
```suggestion
		\add_filter( 'option_woocommerce_subscriptions_enable_retry', [ __CLASS__, 'force_allow_failed_payment_retry' ] );
```

File: includes/reader-revenue/woocommerce/class-woocommerce-connection.php [hunk 3]
```diff
@@ -45,8 +45,10 @@ public static function init() {
...
 		\add_filter( 'default_option_woocommerce_subscriptions_allow_switching', [ __CLASS__, 'force_allow_subscription_switching' ], 10, 2 );
 		\add_filter( 'default_option_woocommerce_subscriptions_allow_switching_nyp_price', [ __CLASS__, 'force_allow_subscription_switching' ], 10, 2 );
+		\add_filter( 'default_option_woocommerce_subscriptions_enable_retry', [ __CLASS__, 'force_allow_failed_payment_retry' ], 10 );
```
Comment (leogermani) [inline includes/reader-revenue/woocommerce/class-woocommerce-connection.php:51]:
```suggestion
		\add_filter( 'default_option_woocommerce_subscriptions_enable_retry', [ __CLASS__, 'force_allow_failed_payment_retry' ] );
```

**Review by leogermani — APPROVED** [review]
Working well. Left two minor suggestions

**Comment by dkoo** [comment]
Thanks for the incremental line-by-line updates, @leogermani. I must be trying to do too many things at once

**Pull request merged**